docs/plutus-report/scripting.tex

```diff
@@ -111,7 +115,127 @@ \subsection{Recursive values}
 
 \subsection{Builtin types and values}
 
-\todompj{Roman is drafting some material for this section}
+Programming languages commonly have some form of foreign function interface. The idea is that:
+
+\begin{enumerate}
+\item
+  We don't want to reimplement everything from the bottom up in a newly designed language.
+  It makes much more sense to delegate certain computations to an existing tool where such computations already exist in optimized form, with good tests etc.
+\item
+  In particular, We certainly don't want to implement our own version of arithmetic (integers, addition, multiplication, etc.) or any other low-level stuff where it will be especially hard to beat external implementations.
+\end{enumerate}
+
+We certainly need efficient primitives in \gls{plutus-core}, including arithmetic.
+
+To implement this, we could build \code{Integer} into the grammar of \gls{plutus-core} types, and \code{Integer} constants into the grammar of \gls{plutus-core} terms, as well as a few functions (addition, multiplication, etc.).
+If we need more types or functions, then we simply have to extend the hardcoded list of types and functions.
+
+However, this is very inflexible.
+The key issue is that we may want to have \emph{different} sets of builtins in different situations.
+For example:
+\begin{enumerate}
+\item
+  During testing it can be convenient to have a builtin which implements \emph{tracing}: emitting log lines to a global trace output.
+  But we don't want this functionality on-chain, so we want to remove the builtin (or replace it with a no-op).
+\item
+  We may need different builtins for different ledgers (see \cref{req:script-lang-extensibility}).
+\end{enumerate}
+
+So we need to make builtin types and functions extensible.
+
+\subsubsection{Builtin types}
+
+Extensible builtin types are relatively straightforward on the type level, but become trickier on the value level.
+Values of builtin types are \emph{constants}, but what is a constant?
+It's easy to say what a \code{Integer} constant should be: an integer!
+However, what about constants of extensible builtin types?
+
+We adopt an approach familiar from generic programming in dependently-typed languages, where we parameterize our language over a \emph{universe} of types.\footnote{
```

Comment:
It's... well, the "universe" term that is usually used there is mostly not for parameterization purposes, but just to refer to all of the permitted types or a subset of them. Then you can do generic things, but only over a particular universe of types. While we're generic over the choice of a universe.